### ISSUE TITLE
create_launch_template breaks cluster DNS

### ISSUE THREAD

**Issue by vprus** [aws-ia/terraform-aws-eks-blueprints#962]
I have created an EKS cluster using the TF file at https://gist.github.com/vprus/b10a5d12f77abbaeba69650b0ca7b3a5.  It creates 
- Managed node group without create_launch_template; coredns gets scheduled there
- Managed node group with create_launch_template

Workload on the second group is unable to resolve cluster DNS names.

The problem seems to be with security groups. I have these security groups
- sg-0d423b7cc90f895d5 / eks-cluster-sg-analytics-2-803067819 (created by EKS, allows ingress from itself)
- sg-02cb5e34ce1e8625a / analytics-2-cluster-20220830085301105500000002 (created by terraform, customizable with cluster_security_group_additional_rules)
- sg-0cdccc699139076f9 / analytics-2-node-20220830085301103400000001 (created by terraform, customizable with node_security_group_additional_rules)

I then observe that
- node in the first nodegroup is associated with eks-cluster-sg-analytics-2-803067819 and only it (both the node and interfaces)
- node in the second nodegroup  is associated analytics-2-node-20220830085301103400000001

eks-cluster-sg-analytics-2-803067819 does not allow ingress from analytics-2-node-20220830085301103400000001, so DNS queries fail. Since node is not associated with analytics-2-cluster-20220830085301105500000002, I can't add additional rules in TF. I can add additional rules to eks-cluster-sg-analytics-2-803067819, and it fixes the problem, but I'd rather avoid manual tweaks.

Is there anything wrong in my TF file, or is this a bug?

**Comment by vprus** [comment]
FOAD, this is the test pod I use to test DNS
```
apiVersion: v1
kind: Pod
metadata:
  name: dnsutils
  namespace: kafka
spec:
  tolerations:
    - key: "purpose"
      operator: "Equal"
      value: "infra"
      effect: "NoSchedule"
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
          - matchExpressions:
            - key: purpose
              operator: In
              values:
              - infra
  containers:
  - name: dnsutils
    image: registry.k8s.io/e2e-test-images/jessie-dnsutils:1.3
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
  restartPolicy: Always
```

**Comment by vprus** [comment]
I have examined launch configs that works and that don't. I appears the broken one has two differences
- the "metadata" field is set to "enabled"
- the "metadata version" field is set to "V2"
If I manually set those two to "not specified" and do instance refresh, DNS starts working.

**Comment by vprus** [comment]
I appears that if I find the node where corends pods are running, and modify its security group to allow ingress from anywhere, DNS starts working. 

The nodegroup where coredns ends up was defined like this:

```
    mg_5 = {
      node_group_name = "managed-ondemand"
      instance_types  = ["m5.large"]
      desired_size    = 2
      min_size        = 2
      max_size        = 2
      subnet_ids      = module.vpc.private_subnets
    }
```

Looks like nodegroups with create_launch_template can't send DNS queries to usual managed node groups, by default?

If so, maybe it should be documented?

**Comment by bryantbiggs** [comment]
when you use the default launch template created by the EKS managed node group service, it will attach the cluster's primary security group. when you use the custom launch template, it will attach the security groups created by the module. to ensure consistency, you should be consistent with your selection of using the default launch template or custom launch template

**Comment by bryantbiggs** [comment]
> That means, that if I started to use create_launch_template=true for any nodegroup, I should switch other existing node groups too? 

That depends on the outcome you are trying to achieve. The guidance was more general in that if you want your node groups to use consistent set(s) of security groups, you will need to configure your code in such a way to match that expectation.

> I guess it's an important constraint that could be added to the documentation, then?

It is documented https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html

**Comment by vprus** [comment]
I'd respectfully disagree. That documentation is the general EKS documentation. The fact that using create_launch_template =true in EKS Blueprints makes the launch template use a different security group that cannot communicate with other node group is behaviour specific to EKS Blueprints, and I think it violates the principle of least surprise, so better be documented.

**Comment by bryantbiggs** [comment]
This is not specific to EKS Blueprints,  it is the behavior of the EKS Managed Node Group service and that is why I linked the appropriate documentation

**Comment by vprus** [comment]
Isn't it possible for EKS Blueprints, when creating launch template, to use cluster's primary security group by default?

**Comment by bryantbiggs** [comment]
It could do that, yes, but it currently does not make that assumption. When users elect to use a custom launch template, it is left up to users to prescribe their intended settings via the custom launch template.

tl;dr 
- Using default launch template means EKS managed node group provides assumptions via default and/or prescriptive settings (i.e. - use cluster primary security group)
- Using custom launch template means users provide their own settings and the module prescribes only the bare minimum settings to successfully create a cluster

**Comment by bryantbiggs** [comment]
closing out with guidance provided above - any further questions, please feel free to open a new issue
